### PR TITLE
Metadata for consumers

### PR DESCRIPTION
**Pull request by mensfeld**
close https://github.com/karafka/karafka/issues/392
close https://github.com/karafka/karafka/issues/335

### CODE REVIEW

File: lib/karafka/connection/batch_delegator.rb [hunk 1]
```diff
@@ -23,14 +23,24 @@ def call(group_id, kafka_batch)
             consumer: consumer,
             kafka_batch: kafka_batch
           ) do
+
```
Comment (nijikon) [inline lib/karafka/connection/batch_delegator.rb]: This line is redundant

File: lib/karafka/params/builders/metadata.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        class << self
+          # Creates metadata based on the kafka batch data
+          # @param kafka_batch [Kafka::FetchedBatch] kafka batch details
+          # @param topic [Karafka::Routing::Topic] topic for which we've fetched the batch
+          # @return [Karafka::Params::Metadata] metadata object
+          def from_kafka_batch(kafka_batch, topic)
+            Karafka::Params::Metadata
+              .new
+              .merge!(
+                'topic' => topic.name,
```
Comment (nijikon) [inline lib/karafka/params/builders/metadata.rb]: Can we sort here?

File: lib/karafka/params/builders/params.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      module Params
+        class << self
+          # @param [Kafka::FetchedMessage] message fetched from Kafka
+          # @param [Karafka::Routing::Topic] topic for which this message was fetched
+          # @return [Karafka::Params::Params] params object
+          def from_kafka_message(message, topic)
+            Karafka::Params::Params
+              .new
+              .merge!(
+                'value' => message.value,
```
Comment (nijikon) [inline lib/karafka/params/builders/params.rb]: Can we sort keys here?

File: lib/karafka/params/metadata.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+module Karafka
+  module Params
+    # Simple metadata object that stores all non-message information received from Kafka cluster
+    # while fetching the data
+    class Metadata < Hash
+      # Attributes that should be accessible as methods as well (not only hash)
+      METHOD_ATTRIBUTES = %w[
```
Comment (nijikon) [inline lib/karafka/params/metadata.rb]: Can we sort keys here?

File: spec/lib/karafka/consumers/includer_spec.rb [hunk 5]
```diff
@@ -8,12 +8,12 @@
...
-      backend: backend,
-      batch_consuming: batch_consuming,
-      responder: responder
-    )
+    build(:routing_topic)
+      .tap do |topic|
+        topic.backend = backend
+        topic.batch_consuming = batch_consuming
+        topic.responder = responder
+      end
```
Comment (nijikon) [inline spec/lib/karafka/consumers/includer_spec.rb]: Can this be?
```ruby
build(
  :routing_topic,
  backend: backend,
  batch_consuming: batch_consuming,
  responder: responder
)
```

File: spec/lib/karafka/params/builders/metadata_spec.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+RSpec.describe Karafka::Params::Builders::Metadata do
+  let(:routing_topic) { build(:routing_topic) }
+  let(:kafka_batch) { build(:kafka_fetched_batch) }
+
+  describe '#from_kafka_batch' do
+    subject(:result) { described_class.from_kafka_batch(kafka_batch, routing_topic) }
+
+    it { is_expected.to be_a(Karafka::Params::Metadata) }
+    it { expect(result.batch_size).to eq kafka_batch.messages.count }
```
Comment (nijikon) [inline spec/lib/karafka/params/builders/metadata_spec.rb]: Can we apply brackets to all? `eq(kafka_batch.messages.count)`

File: spec/lib/karafka/params/builders/params_spec.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+  describe '#from_kafka_message' do
+    subject(:result) { described_class.from_kafka_message(fetched_message, routing_topic) }
+
+    before { Timecop.freeze }
+
+    after { Timecop.return }
+
+    it { is_expected.to be_a(Karafka::Params::Params) }
+    it { expect(result.value).to eq fetched_message.value }
```
Comment (nijikon) [inline spec/lib/karafka/params/builders/params_spec.rb]: Can we apply brackets to all?

File: spec/lib/karafka/params/builders/metadata_spec.rb (hunk 6 above)
Comment (mensfeld) [reply spec/lib/karafka/params/builders/metadata_spec.rb]: there's no rule for that now and for most of the time we don;t use them

Comment (nijikon) [reply spec/lib/karafka/params/builders/metadata_spec.rb]: I will prepare a rule for `rubocop-rspec` as discussed.

**Pull request merged**